docs/enhancements/image-handling-service.md

```diff
@@ -0,0 +1,171 @@
+application. Splitting these responsibilities would allow us to manage and scale
+these deployments separately.
+
+### Goals
+
+* Allow image serving to be scaled independently of the rest of the application.
+* Remove performance impact due to generating and serving images from the assisted service.
+* Make minimal changes to existing (or proposed) APIs.
+
+### Non-Goals
+
+* Serve a discovery image directly to a BMC
+
+## Proposal
+
+The new service (image service) will handle customizing and serving the customized
+image to users. It will query assisted service to retrieve the information that needs
+to be added to the image based on the information passed in the request by the user.
+
+The assisted service will retain the `POST /clusters/{cluster_id}/downloads/image`
+endpoint which will save the image information to its database. In the case
+of the kube API, the controllers will continue to save image information from the
+various CRs as they do currently.
+
+The image service will expose a single API endpoint to download an image. Initially
+this should be compatible with the existing `GET /clusters/{cluster_id}/downloads/image`
+assisted service endpoint. When a user makes a request to this endpoint the
+image service will fetch the ignition and ramdisk (if necessary) from the
+assisted service and stream the customized iso to the client using the
+appropriate template image as a base. This will involve adding an endpoint to
+assisted service to expose the ramdisk data.
+
+### User Stories
+
+#### Story 1
+
+As a user deploying and running assisted service, I need image serving to be
+horizontally scalable and to not impact performance of the assisted service API.
+
+#### Story 2
+
+An additional deployment and service will be managed for all deployment methods.
+This includes SaaS, operator, and CI (minikube). Whether an additional image is
+created is still an open question.
+
+### Implementation Details/Notes/Constraints
+
+#### Authentication and Authorization
+
+The image service will not need to implement authentication or authorization
+directly as it doesn't manage any user data. The user will provide a token to
+the image service and the image service will directly pass that token to
+assisted service on each call. Assisted service will validate the token before
+giving the image service any user information.
+
+#### Template Images
+
+Currently, when the assisted service starts, it downloads and caches each RHCOS
+image specified in the `OPENSHIFT_VERSIONS` environment variable. It also
+creates a minimal iso image for each version by removing the root filesystem.
+These "template" images are the base for creating the discovery iso.
+
+The image service will take on the responsibility of managing the RHCOS template
+images and creating the minimal ISO template images. This means that the image
+service will require persistent storage (when not using S3).
```

Comment:
> How much storage are we talking about

I think it would be around 1GB per rhcos version (minimal and full iso templates).

> The performance benefit from using S3 may just be too substantial, but it's worth thinking about.

I'm not sure that it's a performance gain. Are there any reasons to use S3 in the cloud if we can just re-download the data?
We currently don't use any PVs in the cloud deployment, is that a requirement or could we move the image service to local storage (or ephemeral)?

Reply:
I understand that assisted-service will continue to us S3 to store user files and logs, but I'm not sure it's necessary specifically for storing the template images.